File: app/src/main/java/org/ole/planet/myplanet/service/TaskNotificationWorker.java
package org.ole.planet.myplanet.service;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.model.RealmTeamTask;
import org.ole.planet.myplanet.model.RealmUserModel;
import org.ole.planet.myplanet.utilities.NotificationUtil;
import org.ole.planet.myplanet.utilities.TimeUtils;

import java.util.Calendar;
import java.util.List;

import io.realm.Realm;

public class TaskNotificationWorker extends Worker {
    private Context context;

    public TaskNotificationWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
        this.context = context;
    }

    @NonNull
    @Override
    public Result doWork() {
        Realm mRealm = new DatabaseService(context).getRealmInstance();
        long current = Calendar.getInstance().getTimeInMillis();
        Calendar tomorrow = Calendar.getInstance();
        tomorrow.add(Calendar.DAY_OF_YEAR, 1);
        RealmUserModel user = new UserProfileDbHandler(context).getUserModel();
        if (user != null) {
            List<RealmTeamTask> tasks = mRealm.where(RealmTeamTask.class)
                    .equalTo("completed", false)
                    .equalTo("assignee", user.getId())
                    .equalTo("notified", false)
                    .between("deadline", current, tomorrow.getTimeInMillis())
                    .findAll();

            mRealm.beginTransaction();
            for (RealmTeamTask in : tasks) {
                NotificationUtil.create(context, R.drawable.ole_logo, in.title, "Task expires on " + TimeUtils.formatDate(in.deadline));
                in.isNotified = true;
            }
            mRealm.commitTransaction();
        }

        return Result.success();
    }
}

File: app/src/main/java/org/ole/planet/myplanet/service/TaskNotificationWorker.kt
package org.ole.planet.myplanet.service

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import dagger.hilt.android.EntryPointAccessors
import java.util.Calendar
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.di.WorkerDependenciesEntryPoint
import org.ole.planet.myplanet.utilities.NotificationUtils.create
import org.ole.planet.myplanet.utilities.TimeUtils.formatDate

class TaskNotificationWorker(appContext: Context, workerParams: WorkerParameters) :
    CoroutineWorker(appContext, workerParams) {

    override suspend fun doWork(): Result {
        val current = Calendar.getInstance().timeInMillis
        val tomorrow = Calendar.getInstance()
        tomorrow.add(Calendar.DAY_OF_YEAR, 1)

        val entryPoint = EntryPointAccessors.fromApplication(
            applicationContext,
            WorkerDependenciesEntryPoint::class.java
        )
        val userSessionManager = entryPoint.userSessionManager()
        val teamsRepository = entryPoint.teamsRepository()

        val user = userSessionManager.userModel
        val userId = user?.id
        if (!userId.isNullOrBlank()) {
            val tasks = runCatching {
                teamsRepository.getPendingTasksForUser(userId, current, tomorrow.timeInMillis)
            }.getOrElse { emptyList() }

            if (tasks.isNotEmpty()) {
                tasks.forEach { task ->
                    create(
                        applicationContext,
                        R.drawable.ole_logo,
                        task.title,
                        "Task expires on " + formatDate(task.deadline, ""),
                    )
                }

                val taskIds = tasks.mapNotNull { it.id }.filter { it.isNotBlank() }
                if (taskIds.isNotEmpty()) {
                    runCatching { teamsRepository.markTasksNotified(taskIds) }
                }
            }
        }
        return Result.success()
    }
}
